ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilder.java

```diff
@@ -508,8 +514,19 @@ public GenericRow transform(final K key, final GenericRow row) {
           final long timestamp = processorContext.timestamp();
           final Collection<?> keyColumns = keyGenerator.apply(key);
 
-          row.ensureAdditionalCapacity(1 + keyColumns.size());
+          final int additionalColumns = SystemColumns
+              .getPseudoColumnsFromVersion(pseudoColumnVersion).size();
+
+          row.ensureAdditionalCapacity(additionalColumns + keyColumns.size());
           row.append(timestamp);
+
+          if (pseudoColumnVersion > 0) {
```

Comment:
See my comment above on how to make this more readable (by avoiding magic numbers).